### PR TITLE
Closes #65 | Create dataset loader for filipino-hatespeech-text

### LINKED ISSUE #65: Create dataset loader for Filipino Hate Speech Tiktok - Text
Dataloader name: `filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?filipino_hatespeech_tiktok

|     Dataset| filipino_hatespeech_tiktok |
|-------------|---|
| Description | The dataset contains annotated hate speech from transcribed Tiktok videos, mostly in Taglish (codemixed Tagalog and Cebuano) collected via an unofficial Tiktok API. Most of the domain and context of the videos are related to politics and general elections in the Philippines.  Labeling: 0 - no hate speech, 1 - recognized hate speech |
| Subsets | - |
| Languages | tgl, eng |
| Tasks | Hate Speech Detection |
| License | Creative Commons Attribution Share Alike 4.0 (cc-by-sa-4.0) |
| Homepage | https://github.com/imperialite/filipino-tiktok-hatespeech/ |
| HF URL | - |
| Paper URL | https://dl.acm.org/doi/abs/10.1145/3485768.3485806 |

### PR DESCRIPTION
**Pull request by williamnixon20**
Closes #65 

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+@inproceedings{hernandez2021bert,
+  title={A bert-based hate speech classifier from transcribed online short-form videos},
+  author={Hernandez Urbano Jr, Rommel and Uy Ajero, Jeffrey and Legaspi Angeles, Angelic and Hacar Quintos, Maria Nikki and Regalado Imperial, Joseph Marvin and Llabanes Rodriguez, Ramon},
+  booktitle={2021 5th International Conference on E-Society, E-Education and E-Technology},
+  year={2021}
+}
+"""
+
+_LOCAL = False
+_LANGUAGES = ["tgl"]
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:21]: You can check the related issue https://github.com/SEACrowd/seacrowd-datahub/issues/65 for some of the metadata information.

```suggestion
_LANGUAGES = ["tgl", "eng"]
```

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+_DATASETNAME = "filipino_hatespeech_tiktok"
+
+_DESCRIPTION = """\
+The dataset contains annotated hate speech from transcribed Tiktok videos, mostly in Taglish (codemixed Tagalog and Cebuano) collected via an unofficial Tiktok API. Most of the domain and context of the videos are related to politics and general elections in the Philippines. Labeling: 0 - no hate speech, 1 - recognized hate speech
+"""
+
+_HOMEPAGE = "https://github.com/imperialite/filipino-tiktok-hatespeech"
+_LICENSE = Licenses.UNKNOWN.value
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:29]: I think this is CC_BY_SA_4.0 (c.f. https://github.com/SEACrowd/seacrowd-datahub/issues/65)

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+
+class FilipinoHateSpeechTikTok(datasets.GeneratorBasedBuilder):
+
+    SOURCE_VERSION = datasets.Version(_SOURCE_VERSION)
+    SEACROWD_VERSION = datasets.Version(_SEACROWD_VERSION)
+
+    BUILDER_CONFIGS = [
+        SEACrowdConfig(
+            name="filipino_hatespeech_tiktok_source",
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:46]: Not required, but you can use `_DATASETNAME` so that we don't hard-code every time and avoid potential misspelling errors:
```suggestion
            name="{_DATASETNAME}_source",
```

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+_HOMEPAGE = "https://github.com/imperialite/filipino-tiktok-hatespeech"
+_LICENSE = Licenses.UNKNOWN.value
+_SUPPORTED_TASKS = [Tasks.SENTIMENT_ANALYSIS]
+_SOURCE_VERSION = "1.0.0"
+_SEACROWD_VERSION = "1.0.0"
+
+_URLS = {
+    "raw": "https://raw.githubusercontent.com/imperialite/filipino-tiktok-hatespeech/main/data",
+}
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:36]: Maybe we can use this data structure to its full extent? Either
```suggestion
_URL = "https://raw.githubusercontent.com/imperialite/filipino-tiktok-hatespeech/main/data",
```

or

```suggestion
_URLS = {
    "train": "https://raw.githubusercontent.com/imperialite/filipino-tiktok-hatespeech/main/data/train.csv",
    "valid": "https://raw.githubusercontent.com/imperialite/filipino-tiktok-hatespeech/main/data/valid.csv",
    "test": "https://raw.githubusercontent.com/imperialite/filipino-tiktok-hatespeech/main/data/test.csv",
}
```

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "filipino_tiktok_hatespeech_source"
+
+    def _info(self) -> datasets.DatasetInfo:
+        if self.config.schema == "source":
+            features = datasets.Features({"text": datasets.Value("string"), "label": datasets.Value("string")})
+        elif self.config.schema == "seacrowd_text":
+            features = schemas.text.features()
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:67]: By default, the[ class features here](https://github.com/williamnixon20/seacrowd-datahub/blob/4a8300b136a96494f65dcb1cfdc010f6255f773d/seacrowd/utils/schemas/text.py) are `"Yes"` and `"No"`. We want to be consistent with what's in the dataset, so I suggest passing "0" and "1" here:

```suggestion
            features = schemas.text.features(label_names = ["0", "1"])
```

You can also create a constant, `CLASS_LABELS` and pass that instead.

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        )
+
+    def _split_generators(self, dl_manager: datasets.DownloadManager) -> List[datasets.SplitGenerator]:
+        base_dir = _URLS["raw"]
+        data_files = {
+            "train": base_dir + "/train.csv",
+            "test": base_dir + "/test.csv",
+            "validation": base_dir + "/valid.csv",
+        }
+
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:84]: Not required, but it may be better to be explicit in the download step here, so that the typing is correct. We can add a line with something like:

```python
train_filepath = Path(dl_manager.download_and_extract(_URLS["train"]))
```

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        ]
+
+    def _generate_examples(self, filepath: Path, split: str) -> Tuple[int, Dict]:
+        df = pd.read_csv(filepath).reset_index()
+        id = 0
+        if self.config.schema == "source":
+            for row in df.itertuples():
+                ex = {
+                    "text": row.text,
+                    "label": row.label if not pd.isna(row.label) else 0,
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:116]: This value becomes a float when I run the test script (`0.0` instead of `0`, `1.0` instead of `1`). I suggest turning this into an integer and then typecasting to string. 

```
Generating train split: 10038 examples [00:00, 38879.54 examples/s]
Generating test split: 4254 examples [00:00, 29701.82 examples/s]
Generating validation split: 4246 examples [00:00, 30438.29 examples/s]

INFO:__main__:Dataset sample [source]
{'text': 'Inaasahan na ni Vice President Jejomar Binay na may mga taong... https://t.co/SDytgbWiLh', 'label': '0.0'}
INFO:__main__:Dataset sample [seacrowd_text]
{'id': '0', 'text': 'Inaasahan na ni Vice President Jejomar Binay na may mga taong... https://t.co/SDytgbWiLh', 'label': 0}
INFO:__main__:Checking global ID uniqueness
```

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                }
+                yield id, ex
+                id += 1
+
+        elif self.config.schema == "seacrowd_text":
+            for row in df.itertuples():
+                ex = {
+                    "id": id,
+                    "text": row.text,
+                    "label": row.label if not pd.isna(row.label) else 0,
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:126]: Same comment here. When running the text it seems that it's a full numerical integer instead of a string. I suggest typecasting this fully into a string so that it's consistent.

**Review by ljvmiranda921 — CHANGES_REQUESTED** [review]
Hi @williamnixon20 , thank you so much for the PR!

The data loader runs as intended, but I have a few comments and suggestions to make the implementation more consistent with the other data loaders. I also suggest double-checking the original issue #65 to ensure that the metadata (e.g., License, Languages, etc.) are the same.

Finally, I also suggest running `make check_file=...` to ensure that the source code is formatted properly.

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py (hunk 7 above)
Comment (williamnixon20) [reply seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:116]: image
Thanks for the review! I typecasted the labels to string now

**Review by ljvmiranda921 — APPROVED** [review]
LGTM, thanks for this work :) Let's just wait for @gentaiscool 's comments.

**Comment by holylovenia** [comment]
I'm re-assigning this dataloader's review task to @sabilmakbar since @gentaiscool currently has his hands full with something else.

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+_HOMEPAGE = "https://github.com/imperialite/filipino-tiktok-hatespeech"
+_LICENSE = Licenses.CC_BY_SA_4_0.value
+_SUPPORTED_TASKS = [Tasks.SENTIMENT_ANALYSIS]
+_SOURCE_VERSION = "1.0.0"
+_SEACROWD_VERSION = "1.0.0"
+
+_URL = "https://raw.githubusercontent.com/imperialite/filipino-tiktok-hatespeech/main/data"
+
+
+class FilipinoHateSpeechTikTok(datasets.GeneratorBasedBuilder):
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:38]: pls add `Dataset` suffix on the class name, for conformity
```suggestion
class FilipinoHateSpeechTikTokDataset(datasets.GeneratorBasedBuilder):
```

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        ]
+
+    def _generate_examples(self, filepath: Path, split: str) -> Tuple[int, Dict]:
+        df = pd.read_csv(filepath).reset_index()
+        id = 0
+        if self.config.schema == "source":
+            for row in df.itertuples():
+                ex = {
+                    "text": row.text,
+                    "label": str(int(row.label)) if not pd.isna(row.label) else "0",
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py:114]: Is this behavior also used on their original dataset? Or is it being used only for convenience?

if yes, I'd like to suggest adding a reference link to it; else, you may skip it on iteration so that we don't alter the label/rule from original implementation

**Review by sabilmakbar — CHANGES_REQUESTED** [review]
Initial comments, will try to dry-run this dataset later, eta by EoD UTC+7

File: seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py (hunk 10 above)
Comment (williamnixon20) [reply seacrowd/sea_datasets/filipino_hatespeech_tiktok/filipino_hatespeech_tiktok.py]: Hi, I've updated it, thanks! :D

**Comment by sabilmakbar** [comment]
The dataloader works as intended, but I'd argue to keep NA datapoints on `source` implementation (we can skip it on SEACrowd Schema, though). What's your take, LJ @ljvmiranda921, given the details of NA datapoints shared below?

## Details
I've checked their dataset, here's the label distribution among the splits:
Dset Info

In addition, here's the sample of the null data among three different splits:
1. `valid.csv`
Valid NA Example

2. `test.csv`
Test NA Example

3. `train.csv`
Train NA Example

**Comment by ljvmiranda921** [comment]
Hi @sabilmakbar , thanks for this! 

>  but I'd argue to keep NA datapoints on source implementation so it will match with original dataset (we can skip it on SEACrowd Schema, though)

For the `source` implementation, you're right, we should probably maintain their parity with the original implementation, and then skip it in the SEACrowd schema.

Looking at the datapoints, the `NA` labels don't look very informative. Most of them are hashtags that do not clearly fall between the original labels. I'm not sure if they're intended, it might even be parsing errors. But yeah, let's keep them in the `source` implementation. I haven't read the paper fully, so the authors might have used them in some experiment somewhere. 

Thanks for checking @sabilmakbar and for the catch @williamnixon20 !

**Comment by sabilmakbar** [comment]
> For the source implementation, you're right, we should probably maintain their parity with the original implementation, and then skip it in the SEACrowd schema.
> 
> Looking at the datapoints, the NA labels don't look very informative. Most of them are hashtags that do not clearly fall between the original labels. I'm not sure if they're intended, it might even be parsing errors. But yeah, let's keep them in the source implementation. I haven't read the paper fully, so the authors might have used them in some experiment somewhere.

Thanks for your input, LJ! I'll approve this now as the most recent implementation is already aligned with our latest decision on keeping data as is in `source` schema.

**Review by sabilmakbar — APPROVED** [review]
LGTM! Thanks for the hard work, @williamnixon20!

**Comment by sabilmakbar** [comment]
I'm merging this...
